### PR TITLE
cmd/signer: Add extract command, add entity signer pk check

### PR DESCRIPTION
**Pull request by david-yan**
Resolves https://github.com/oasislabs/oasis-core/issues/2609

- Adds `oasis-node signer extract` command
- Adds check to entity and signer public keys when running `entity.Load`

### CODE REVIEW

File: go/oasis-node/cmd/signer/signer.go [hunk 1]
```diff
@@ -12,14 +17,44 @@ var (
 		Use:   "signer",
 		Short: "signer backend utilities",
 	}
+
+	extractCmd = &cobra.Command{
+		Use:   "extract",
```
Comment (ravenac95) [inline go/oasis-node/cmd/signer/signer.go]: can we make this `export-public`?

File: go/oasis-node/cmd/common/common.go [hunk 2]
```diff
@@ -278,3 +278,17 @@ func LoadEntity(signerBackend string, entityDir string) (*entity.Entity, signatu
 
 	return entity.Load(entityDir, factory)
 }
+
+func ExtractEntity(signerBackend string, entityDir string) error {
```
Comment (kostko) [inline go/oasis-node/cmd/common/common.go]: Add documentation comment, consider renaming (also below you call this exporting) as this just creates an empty entity from a public key?

File: go/common/entity/entity.go [hunk 3]
```diff
@@ -77,6 +77,11 @@ func Load(baseDir string, signerFactory signature.SignerFactory) (*Entity, signa
 		return nil, nil, err
 	}
 
+	if ent.ID != signer.Public() {
+		signer.Reset()
+		return nil, nil, fmt.Errorf("signer, entity public key mismatch")
```
Comment (kostko) [inline go/common/entity/entity.go]: Show the actual keys, e.g., `(entity: %s signer: %s)`.

File: .changelog/2609.feature.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Export Ledger entities
```
Comment (kostko) [inline .changelog/2609.feature.md]: This doesn't seem ledger specific? I would rather says "Export signer public keys as empty entities" or something like that.

File: go/oasis-node/cmd/signer/signer.go [hunk 5]
```diff
@@ -12,14 +17,44 @@ var (
 		Use:   "signer",
 		Short: "signer backend utilities",
 	}
+
+	exportCmd = &cobra.Command{
+		Use:   "export",
+		Short: "export the public key from signer to an entity",
```
Comment (kostko) [inline go/oasis-node/cmd/signer/signer.go]:
```suggestion
		Short: "export the public key from signer as an empty entity",
```

File: go/common/entity/entity.go [hunk 6]
```diff
@@ -77,6 +77,11 @@ func Load(baseDir string, signerFactory signature.SignerFactory) (*Entity, signa
 		return nil, nil, err
 	}
 
+	if ent.ID != signer.Public() {
+		signer.Reset()
+		return nil, nil, fmt.Errorf("public key mismatch\nsigner: %s\nentity: %s", signer.Public(), ent.ID)
```
Comment (kostko) [inline go/common/entity/entity.go]: Don't use newlines in error messages, do something like:
```suggestion
		return nil, nil, fmt.Errorf("public key mismatch (signer: %s entity: %s)", signer.Public(), ent.ID)
```

File: go/common/entity/entity.go [hunk 7]
```diff
@@ -77,6 +77,11 @@ func Load(baseDir string, signerFactory signature.SignerFactory) (*Entity, signa
 		return nil, nil, err
 	}
 
+	if ent.ID != signer.Public() {
```
Comment (kostko) [inline go/common/entity/entity.go]: While this works as the current implementation is a fixed-size array, please use `!ent.ID.Equal(signer.Public())` to be consistent with other parts.

File: go/oasis-node/cmd/signer/signer.go [hunk 8]
```diff
@@ -12,14 +17,44 @@ var (
...
+	}
+	entityDir, err := cmdFlags.SignerDirOrPwd()
+	if err != nil {
+		logger.Error("failed to retrieve signer dir",
+			"err", err,
+		)
+		os.Exit(1)
+	}
+	if err := cmdCommon.ExportEntity(cmdFlags.Signer(), entityDir); err != nil {
+		logger.Error("failed to extract entity",
```
Comment (kostko) [inline go/oasis-node/cmd/signer/signer.go]:
```suggestion
		logger.Error("failed to export entity",
```

**Comment by david-yan** [comment]
@kostko should all be done. Sorry for not squashing the commit message properly in my previous PR.

File: go/common/entity/entity.go (hunk 6 above)
Comment (kostko) [reply go/common/entity/entity.go]: Hm, this still uses newlines?

Comment (david-yan) [reply go/common/entity/entity.go]: Oh. How embarrassing. I thought I committed this suggestion.

**Pull request merged**